### DISCUSSION TITLE
Invalidate ISR function, like res.revalidate but without calling API, only invalidate cache.

### DISCUSSION THREAD

**Discussion by IgorZanellaDev** [vercel/next.js#46654] [Ideas]
### Describe the feature you'd like to request

I have built a multitenant project on NextJS with ISR; some pages are SSR. We have more than 2k websites, and we use middleware to manage the domains. We currently have 20k pages built with ISG.
Sometimes we have to rebuild one of the websites completely, but obviously, we can't relaunch the complete 20k pages build. We also can't launch 200 revalidate requests because it's not suitable for API.
I would like to have a function to only invalidate the page without calling the API instantly but recreate only on the page load like it was invalid from the cache.

### Describe the solution you'd like

There should be a function in response, the same as revalidate. It could be called `res.invalidate(path)`, and it invalidates but does not recreate the static page.

### Describe alternatives you've considered

I tried to do a single /api/revalidate endpoint with multiple revalidate, but we can't do too many API requests.

**Comment by JannikWempe** [comment]
We have the same problem. We'd like to invalidate a potentially huge amount of pages. We can not use on-demand revalidation (at least not for all pages at once) because that would DDoS our own infrastructure. We'd like to just invalidate the cache and have the page updated on the next request (that is just a little slower and blocking while regenerating the page).

## A hack we have tried
We thought we might could purposefully throw an error in `getStaticProps` because [the docs say](https://nextjs.org/docs/pages/building-your-application/data-fetching/incremental-static-regeneration#error-handling-and-revalidation):
> If there is an error inside `getStaticProps` when handling background regeneration, or you manually throw an error, the last successfully generated page will continue to show. On the next subsequent request, Next.js will retry calling `getStaticProps`.

We made this work by setting a flag in a cache indicating that we are revalidating a route and reading & deleting that in `getStaticProps` and throw an error if present. (Would be handy if there would be a way to tell from the `context` that it is a revalidation request).

But it turned out that `getStaticProps` is not hit on the next visit after throwing an error. I am assuming that the "background generation" mentioned in the docs is referring to the first visit after a revalidation period is over?

## Currently no way?
So we are currently not aware of a nice way to serve fresh data to users for a huge amount of pages while utilising ISR with a longer revalidation period. Or are we missing something?

**Comment by DevMude** [comment]
Looks like the app router now has the kind of functionality you're looking for:

https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#on-demand-revalidation
https://nextjs.org/docs/app/api-reference/functions/revalidatePath
